### PR TITLE
Use aggregated discovery API to discovery API groups and resources

### LINKED ISSUE #7526: Use the aggregated discovery to discover the groups/resources
Currently, Velero creates a discovery client when starting up, fetches the API groups every 5 minutes and caches them.
When a backup is created, Velero iterates the cached groups and fetches the resource types for every group and then fetches the objects for every resource type.

Kubernetes introduces a new accept header `Accept: application/json;as=APIGroupDiscoveryList;v=v2beta1;g=apidiscovery.k8s.io` in APIs `api`/`apis` to support returning an aggregated discovery document for all groups/resources in one request.
And Etag is also supported in the new API.
Refer to the [design](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3352-aggregated-discovery/README.md).

With the new machinesm, we can refactor the discovery process to avoid useless interactions with API server and improve the performance for backing up Kuerbernetes resources.

**Vote on this issue!**

This is an invitation to the Velero community to vote on issues, you can see the project's [top voted issues listed here](https://github.com/vmware-tanzu/velero/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).  
Use the "reaction smiley face" up to the right of this comment to vote.

- :+1: for "I would like to see this bug fixed as soon as possible"
- :-1: for "There are more important bugs to focus on right now"

### PR DESCRIPTION
**Pull request by ywk253100**
Use aggregated discovery API to discovery API groups and resources

Fixes #7526

Thank you for contributing to Velero!

# Please add a summary of your change

# Does your change fix a particular issue?

Fixes #(issue)

# Please indicate you've done the following:

- [ ] [Accepted the DCO](https://velero.io/docs/v1.5/code-standards/#dco-sign-off). Commits without the DCO will delay acceptance.
- [ ] [Created a changelog file (`make new-changelog`)](https://velero.io/docs/main/code-standards/#adding-a-changelog) or comment `/kind changelog-not-required` on this PR.
- [ ] Updated the corresponding documentation in `site/content/docs/main`.

### CODE REVIEW

File: pkg/test/discovery_client.go [hunk 1]
```diff
@@ -80,3 +81,11 @@ func (c *DiscoveryClient) WithAPIResource(resource *APIResource) *DiscoveryClien
...
 	return c
 }
+
+func (c *DiscoveryClient) GroupsAndMaybeResources() (*metav1.APIGroupList, map[schema.GroupVersion]*metav1.APIResourceList, map[schema.GroupVersion]error, error) {
+	apiGroupList, err := c.ServerGroups()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return apiGroupList, nil, nil, nil
+}
```
Comment (kaovilai) [inline pkg/test/discovery_client.go:91]: This function isn't referenced anywhere?

Comment (ywk253100) [reply pkg/test/discovery_client.go:91]: This function is required for `DiscoveryClient` as a mock implementation for interface `discovery.AggregatedDiscoveryInterface`.

It's called inside the `client-go/discovery` code

Comment (kaovilai) [reply pkg/test/discovery_client.go:91]: gotcha. makes sense.

**Review by kaovilai — APPROVED** [review]

File: pkg/discovery/helper.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ type serverResourcesInterface interface {
 }
 
 type helper struct {
```
Comment (blackpiglet) [inline pkg/discovery/helper.go:76]: The `pkg/discovery/helper.go` is a porting module from the old version of the client-go package.
Is it possible that we deprecate this module, and use the client-go code instead?

Comment (ywk253100) [reply pkg/discovery/helper.go:76]: Discussed with @blackpiglet offline, the `pkg/discovery/helper.go` calls client-go's discovery code with Velero needed logic wrapped, cannot be replaced by client-go's discovery.

**Comment by ywk253100** [comment]
@blackpiglet @shubham-pampattiwar @Lyndon-Li Could you take a look at this PR again?

**Review by shubham-pampattiwar — APPROVED** [review]

**Comment by shubham-pampattiwar** [comment]
@ywk253100 Are we planning to conduct any performance testing to measure how much the backup operation’s performance improves with the PR changes?

**Review by blackpiglet — APPROVED** [review]

**Review by Lyndon-Li — APPROVED** [review]

**Comment by ywk253100** [comment]
> @ywk253100 Are we planning to conduct any performance testing to measure how much the backup operation’s performance improves with the PR changes?

No plan at this moment.

**Pull request merged**